fennel/datasets/datasets.py

```diff
@@ -764,19 +777,41 @@ def __init__(
         node: _Node,
         keys: List[str],
         type: WindowType,
-        gap: Duration,
         field: str,
+        gap: Optional[Duration],
+        duration: Optional[Duration],
+        stride: Optional[Duration],
     ):
         super().__init__()
         if len(keys) == 0:
             raise ValueError(
                 "'group_by' before 'window' must specify at least one key"
             )
+        if stride is None:
+            if type == WindowType.Hopping:
+                raise ValueError("'hopping window' must specify stride")
+
+        if duration is None:
+            if type == WindowType.Hopping or type == WindowType.Tumbling:
+                raise ValueError(
+                    "'hopping window' or 'tumbling window' must specify duration"
+                )
+
+        if gap is None:
+            if type == WindowType.Sessionize:
+                raise ValueError("'sessionize window' must specify gap")
+
```

Comment:
Nit but I think it will be cleaner if you did validation by first matching on window type. That will allow you to validate that all required parameters are provided AND none of the parameters that don't make sense aren't provided.

Reply:
Could you please add some tests for these scenarios too?